Added pathlib support for several functions

wackywendell opened the pull request:
Added support for pathlib.Path support for several functions, including `savetxt`, `loadtxt`, `load`, `savez`, `savez_compressed`, `ndfromtxt`, `recfromtxt`, and `recfromcsv`, along with testing. Fixes #6418.

Note that #6655 also addresses the same issue, and we should coordinate on this; I was a bit slow on making my changes, and @r0fls did some good work on his branch!

This could also use a bit of code review, I imagine.

CC @rgommers, @r0fls.

wackywendell committed:
ENH: Add support for pathlib.Path objects to save/load functions

r0fls commented on numpy/lib/npyio.py:
```diff
@@ -425,7 +429,7 @@ def save(file, arr, allow_pickle=True, fix_imports=True):
-    file : file or str
+    file : file, str, or pathlib.Path
         File or filename to which the data is saved.  If file is a file-object,
         then the filename is unchanged.  If file is a string, a ``.npy``
```
You could update this to say `If file is a string or a pathlib.Path object, a ''.npy''... extension will be appended to the file name if it does not already...`

wackywendell replied:
Good point, and done!

charris commented on numpy/lib/npyio.py:
```diff
@@ -86,10 +86,12 @@ def __dir__(self):
 
 
-def zipfile_factory(*args, **kwargs):
+def zipfile_factory(file, mode='r', *args, **kwargs):
```
Umm, adding a positional argument seems like a potential compatibility problem.

wackywendell replied:
I'm not sure how in this case it would be a problem. Could you be more specific?

Before, it was just passing `*args` to `zipfile.Zipfile()`; the first two args to  `zipfile.Zipfile()` are necessarily `file` and `mode`, and AFAICT always have been. I'm not sure how this could be a backwards-compatibility problem.

Or are you thinking this might be a *future* compatibility problem?

charris replied:
By and large, signatures should not be changed without good reason, there is always the possibility that you have missed something nor can you rely on the tests being good enough to catch problems. This is particularly true for public functions, which unfortunately this is. The original was perhaps unfortunate, but there you go. It is best to keep PRs focused. If you feel this change is a useful cleanup, then it should come in as a separate PR.

charris replied:
Just to clarify the procedure ;) A new private function, say `_zipfile_factory` would be defined, and `zipfile_factory` would be deprecated. That may seem overly picky, but all such changes carry some risk, and when several hundred PRs are present in a release the risk accumulates.

wackywendell replied:
OK, that makes sense. I'll work on it - probably by changing it back, because I don't think this is important enough to warrant a deprecation cycle.

wackywendell replied:
I reverted the type signature, although I did add an `assert`. With the previous version, a 0-length `*args` would cause an error from `zipfile.Zipfile`; this just changes the exception that would be raised from a `TypeError` to an `AssertionError`. Is that ok, or should I also preserve the assertion type?

charris commented on numpy/lib/_iotools.py:
```diff
@@ -50,6 +54,13 @@ def _is_bytes_like(obj):
+    """
+    Check whether obj is a pathlib.Path object.
+    """
+    return PurePath is not None and isinstance(obj, PurePath)
```
Do you need to check `PurePath is not None`?

charris replied:
`PurePath` looks like the wrong type to check as it may be purely abstract, `Path` would seem better.

wackywendell replied:
Hmm, that is an interesting thought, but I just plugged it into the interpreter and it gives a `TypeError` for `isinstance(x, None)`:

```python
>>> p=Path('x')
>>> isinstance(p, None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: isinstance() arg 2 must be a type or tuple of types
```

Note that `PurePath` will be `None` if pathlib is not installed, so the `PurePath is None` is protecting that `isinstance`.

wackywendell replied:
> `PurePath` looks like the wrong type to check as it may be purely abstract, `Path` would seem better.

Ah, I misread the docs and thought `Path` was an alias for `PosixPath` or `WindowsPath`. You're right, those both inherit from `Path`.

However, it doesn't look to me like `PurePath` is ever abstract, and I'm not sure it matters anyway: I'm only using it for `isinstance`, so its fine if its abstract.

That said, there's no reason to not just go with `Path`, so maybe we should do that.

charris commented on numpy/lib/_iotools.py:
```diff
@@ -50,6 +54,13 @@ def _is_bytes_like(obj):
     return True
 
 
+def _is_path(obj):
```
`is_path` seems too generic for a specific type. Maybe `is_pathlib_path`.

wackywendell replied:
Good point. I'll do that.

charris commented on numpy/lib/_iotools.py:
```diff
@@ -9,6 +9,10 @@
+try:
+    from pathlib import Path, PurePath
+except ImportError:
+    PurePath = Path = None
```
Do you really need `PurePath`?

charris commented:
See also #6655, which does much the same.

wackywendell commented:
> See also #6655, which does much the same.

Yes, #6655 adds `pathlib.Path` support to one function; this adds it to about 7. I could rebase this onto #6655, though, if you'd like... I was waiting to see if #6655 would be merged first, though.

charris commented:
The PR list tends to act like a lifo :) I just saw #6655 as I was going down the list.

charris commented on numpy/lib/npyio.py:
```diff
@@ -87,6 +87,13 @@ def __dir__(self):
 def zipfile_factory(*args, **kwargs):
+    """Create a zipfile.Zipfile. Allows for Zip64, as well as pathlib.Path 
+    objects."""
+    assert len(args) >= 1, "zipfile_factory requires a file"
```
I think you can keep the file argument, it isn't optional in the ZipFile constructor and you would get an error anyway without it.
```
class zipfile.ZipFile(file[, mode[, compression[, allowZip64]]])
```

charris replied:
Note that `assert` doesn't work when python is run in optimized mode. Just raise an error if you want to do it that way. Personally, I think a `TypeError` from `ZipFile` is fine, especially as it is unlikely to ever be seen.

wackywendell replied:
Alright, now its `zipfile_factory(file, *args, **kwargs)`, and the code is cleaner.

charris commented on numpy/lib/npyio.py:
```diff
@@ -476,6 +482,11 @@ def save(file, arr, allow_pickle=True, fix_imports=True):
         own_fid = True
+    elif _is_pathlib_path(file):
+        if not file.name.endswith('.npy'):
+            file = file.parent / (file.name + '.npy')
```
Out of curiosity, what does the `/` do here. I haven't seen that before.

charris replied:
Ah, looked it up.

charris commented on numpy/lib/tests/test_io.py:
```diff
@@ -1823,6 +1828,123 @@ def test_auto_dtype_largeint(self):
 
+
+class TestPathUsage(TestCase):
+    @contextlib.contextmanager
```
This may not handle exceptions correctly, writing a standard context manager might be better.

charris replied:
There is another PR for adding a context manager for temporary files, see #6545 and comments there.

wackywendell replied:
Ah, interesting!

I'm not sure I completely understood the problem, though.

From my understanding, a function like so:

```python
@contextlib.contextmanager
def with_file(fname):
    fd = open(fname)
    yield fd
    fd.close()
```

Will not actually run the `fd.close()` code if an exception is raised.

In this case, though, we have a `yield *inside* another with statement (`with tempdir(prefix=prefix) as tmpdir:`); AFAICT, I think that with statement (`with tempdir…`) will necessarily clean up the tempdir either way. Is that incorrect?

charris replied:
Hmm, if we just supply a name, which I think is the right thing to do, there is nothing that needs closing, that should be handled elsewhere.

wackywendell replied:
I don't understand. The purpose of this context manager is to 1) create a temp directory, 2) get a new temp filename in that directory, and 3) remove that directory when the context ends. What name are we supplying from where to where? How does this help?

charris replied:
I don't really know, it needs some experimentaiton ;) Maybe aim for something like
```
with TempFilePath(...) as fpath:
```

charris replied:
Could probably omit the directory. If I grep `mkstemp` in the repo there are no directories specified and I don't see why we would need one.

wackywendell replied:
It's done [here](https://github.com/numpy/numpy/blob/master/numpy/lib/tests/test_io.py#L318-L325).

My understanding is that if we were to use `mkstemp`, it would only guarantee that that file would be unique *until we removed it*. For this particular test, we need to get a unique filename (done here through `mkstemp`), then *delete* that file, so that we can start with a path with nothing on it. With a bare `mkstemp`, I think its conceivable (if very unlikely) that in between closing it and reopening it the OS might give it out to another program. However, if we ask for a unique *directory*, then we don't need to worry about anything in that directory being reassigned.

Maybe its unnecessary to use both `tempdir` and `mkstemp`, but I'm not sure its 100% safe to do this without the `tempdir` if we need not just an empty file, but a valid path without a file there.

charris replied:
Oh, forgot, we do need to delete the file (and maybe directory) when it finishes, maybe also close it just in case.

wackywendell replied:
OK, we've now said a lot of things, and I lost track a bit of what the current problem is and/or what the solution would be.

My understanding is still that as written, the tempdir would be removed even if an exception were raised. Is that in dispute? If so, why?

Sorry for pushing the issue, its just that I'm not sure I can fix the problem if I don't understand it, because this does involve both a file that may need to be closed as well as a tempdir that needs to be removed, so with that level of complexity I'm still not sure how to proceed without a better understanding of the issue involved.

charris replied:
There are some protections, only the owner can access the file and I would expect other calls to mkstemp would have to return a different name as long as the original exists. There are also suffixes and prefixes that can be specified. The name cannot be reused by another process unless it manages to guess the name, etc, so I expect it is pretty safe. Note some of the usage patterns for mkstemp in the same test file.

For the way forward, probably the easiest thing to do is create a nice context handler in the test file and use that for this PR. Then make another PR moving the manager to `numpy/testing/utils.py` and maybe fixing up some other tests to use it.

juliantaylor replied:
except for our tempdir context manager being broken this looks fine to me
fixing tempdir can be in a separate PR

though one could also go the route of creating a tempdir in the TestCase.setUp and delete it in tearDown, it seems each test only wants one file anyway

charris replied:
There is also a use for temporary files in making zip files, so a context manager would be useful for more than just tests.

charris replied:
| then delete that file

Closing a file created by mkstemp does not delete it, that needs to be done separately.

charris commented on numpy/lib/tests/test_io.py:
```diff
@@ -1823,6 +1828,123 @@ def test_auto_dtype_largeint(self):
+    
+    # Test that pathlib.Path can be used
+    @np.testing.dec.skipif(Path is None,
+    "Cannot test without a pathlib.Path instance")
```
Indent should line up with the `P` in `Path` above. See PEP8.

charris commented on numpy/lib/tests/test_io.py:
```diff
@@ -1823,6 +1828,123 @@ def test_auto_dtype_largeint(self):
+    "Cannot test without a pathlib.Path instance")
+    def test_save_load(self):
+        # Test that pathlib.Path instances can be used with savez.
+        with self.with_path(prefix='numpy_test_path_save_load_', suffix='.npy') as path:
```
Umm, need to keep lines < 80 characters. This looks marginal, although can't tell on screen.

charris replied:
could probably shorten `'numpy_test_path_save_load_'`.

charris commented:
I think we are almost there. Could you squash the commits and take a look at the commit message guidelines in `doc/source/dev/gitwash/development_workflow.rst`. Not sure what to do about the context manager for temporary files, as mentioned there is another PR with one. I'd like to have one available somewhere global, maybe in `numpy/testing/utils.py`

wackywendell commented:
> I think we are almost there. Could you squash the commits and take a look at the commit message guidelines

Thanks! Once we figure out the context manager thing, I'll squash / rebase / fix the commit messages.

rherault-insa commented:
+++ to @charris we definitively need context managers for temp dirs and temp files that can be closed without removed in a more larger context that tests.

rherault-insa commented:
Moreover having functions accepting file descriptor and not only string or file path is a good move in the fact that they will be able to transparently accept in mem (temporary) files.

charris commented:
I've committed a `temppath` context manager in #6866. If you look there, I also refactor a test to use it. You can get at it (and tempdir) by rebasing on master and importing from `numpy.testing`.

wackywendell commented:
Great! I'll look into that.

charris commented:
@wackywendell Could you finish this up and squash the commits into one with a rewritten commit message ala `doc/source/dev/gitwash/development_workflow.rst`?

wwendell commented:
Hi, sorry - I've been a bit side-tracked in the past month or two; I have a PhD defense in a month, and I'm going through interview stuff now, so this is taking me longer than I planned. I'll try and get it done shortly, though.

wackywendell commented:
Hi, sorry - I've been a bit side-tracked in the past month or two; I have a PhD defense in a month, and I'm going through interview stuff now, so this is taking me longer than I planned. I'll try and get it done shortly, though.

charris commented:
Oh, you sound busy ;) OK, I'll put this up to 1.12, but if you don't think you will have time in the near future, let me know and I'll try to finish it up.

wackywendell commented:
Wow, its been a while, but I finally finished this up. It now uses the `numpy.testing.temppath` context manager for testing, and I squashed the commits. Let me know what you think!

As a side note, the pressure is off, so I should be more responsive now.

homu commented:
:umbrella: The latest upstream changes (presumably #7438) made this pull request unmergeable. Please resolve the merge conflicts.

charris commented on numpy/lib/npyio.py:
```diff
@@ -86,10 +86,14 @@ def __dir__(self):
 
-def zipfile_factory(*args, **kwargs):
+def zipfile_factory(file, *args, **kwargs):
+    """Create a zipfile.Zipfile. Allows for Zip64, as well as pathlib.Path
```
Please use standard docstring. Summary line, blank line, extended summary. Here that would be something like
```
"""Create a zipfile.

Extended summary.
"""
```
If you are going to document it, might say more about the arguments.

charris commented on numpy/lib/npyio.py:
```diff
@@ -355,12 +359,14 @@ def load(file, mmap_mode=None, allow_pickle=True, fix_imports=True,
         fid = open(file, "rb")
         own_fid = True
+    elif _is_pathlib_path(file):
+        # e.g., Path instance
```
Don't need the comment, is obvious ;)

charris commented on numpy/lib/tests/test_io.py:
```diff
@@ -1829,6 +1833,115 @@ def test_auto_dtype_largeint(self):
+            assert_array_equal(x, a)
+
+    @np.testing.dec.skipif(Path is None,
+                           "Cannot test without a pathlib.Path instance")
```
Could shorten "Cannot test without a pathlib.Path instance" to "No pathlib.Path". Also below.

charris commented:
Couple of nitpicky comments. I wonder if we should also deal with `memmap`? See `numpy/core/memmap.py`.

wackywendell commented:
Good comments, thanks for the help.

I started work on `memmap` — that seems like another obvious place for this, thanks — but ran into an issue: what do I do about importing `numpy/lib/_iotools/_is_pathlib_path`? I can't import from `lib` into `core`, right? So I could either duplicate the functionality in `core`, or move it to `core` and then also import it into `_iotools`, which seems more right. In that case, where should it go? Maybe in `numpy.compat.py3k`, since `pathlib` is only officially in Python 3?

wackywendell commented:
OK, I added another commit for `memmap`, and moved `_is_pathlib_path` to `numpy.compat.py3k`. Let me know what you think and we can roll that back or squash it.

njsmith commented:
NB that there's a discussion going on on python-ideas right now about what the best approach is to making code compatible with pathlib -- some options under consideration include getting rid of pathlib (!), defining a standard protocol where every function that takes a path does `getattr(obj, "__path__", obj)` without checking the type, making pathlib Paths a subclass of `str` so that no checking is needed, etc.

charris commented:
@njsmith Yeah, I noticed that pathlib was provisional in Python 3.4, makes one a bit nervous...
@wackywendell I probably would have just duplicated the function, it's not big ;)

wackywendell commented:
@charris: OK, do you think I should just duplicate the function? Or is it OK to put it `compat.py3k`?

charris commented:
@wackywendell Not worth worrying about, you can leave it as is.

charris commented:
Although it might be worth removing the leading `_` as it isn't remotely private anymore.

wackywendell commented:
OK, done, done, and rebased! Anything else?

njsmith commented:
Only concern is that it sounds like python-dev is converging on a nicer way to do this that will probably be released in the next 3.4, 3.5 point releases. Maybe we don't care and can always clean this up later once that happens -- I don't have a strong opinion either way.

njsmith commented:
Probably we should just go for it given that the PR exists :-). I won't merge myself though b/c I'm just thinking about the high-level issue, I haven't actually read the code.

charris commented on numpy/compat/py3k.py:
```diff
@@ -10,6 +10,10 @@
            'integer_types']
 
```
Add `Path` and `is_pathlib_path` to the `__all__` list, then you can import from `numpy.compat`.

charris commented on numpy/core/memmap.py:
```diff
@@ -3,6 +3,7 @@
 import numpy as np
 from .numeric import uint8, ndarray, dtype
 from numpy.compat import long, basestring
+from numpy.compat.py3k import is_pathlib_path
```
Should be able to import directly from `numpy.compat` after fix.

charris commented on numpy/core/tests/test_memmap.py:
```diff
@@ -4,6 +4,10 @@
 import os
 import shutil
 from tempfile import NamedTemporaryFile, TemporaryFile, mktemp, mkdtemp
+try:
```
Could import `Path` from `npy_compat`

charris commented on numpy/lib/_iotools.py:
```diff
@@ -9,6 +9,7 @@
 import numpy as np
 import numpy.core.numeric as nx
 from numpy.compat import asbytes, bytes, asbytes_nested, basestring
+from numpy.compat.py3k import is_pathlib_path
```
Best import this from `numpy.compat` so everything is in one place.

charris replied:
In fact, I don't think you need this anymore.

charris commented on numpy/lib/npyio.py:
```diff
@@ -14,8 +14,8 @@
     LineSplitter, NameValidator, StringConverter, ConverterError,
-    ConverterLockError, ConversionWarning, _is_string_like, has_nested_fields,
-    flatten_dtype, easy_dtype, _bytes_to_name
+    ConverterLockError, ConversionWarning, _is_string_like, is_pathlib_path,
```
Might as well import from `npy_compat` so that everything is in one place.

charris commented on numpy/lib/npyio.py:
```diff
@@ -86,10 +86,17 @@ def __dir__(self):
 
 
-def zipfile_factory(*args, **kwargs):
+def zipfile_factory(file, *args, **kwargs):
```
Should be OK, looks like zipfile_factory is effectively private.

wackywendell replied:
This is not changing the effective signature - `*args` was being passed to `zipfile.ZipFile` before, and `zipfile.ZipFile` takes `file` as its first argument, so anything that worked before should continue to work.

charris commented:
I think we are getting close. @njsmith Should be easy to search and fixup. What changes are being proposed?

njsmith commented:
@charris: it looks like the current trend is towards adding a standard convert-to-filesystem-path protocol sorta along the lines of `__index__`, so APIs that take paths and want strings would do `from operators import fspath` and then
```python
def some_file_function(path, ...):
    # force path to be a str
    path = fspath(path)
    ...
```
Or something like that.

Also probably pathlib support will get added to a bunch of builtins like `open`, so that for most code no special handling will be needed at all -- you just write
```
def some_file_function(path, ...):
    with open(path) as f:
        ...
```
and it just magically works for both str and Path arguments.

njsmith commented:
But, it's a mailing list discussion, so -- that's where things seem to be going right now, but, check back next week for our thrilling conclusion :-).

charris commented on numpy/lib/tests/test_io.py:
```diff
@@ -1829,6 +1833,115 @@ def test_auto_dtype_largeint(self):
+
+class TestPathUsage(TestCase):
+    # Test that pathlib.Path can be used
+    @np.testing.dec.skipif(Path is None,
```
Umm, this will now fit on a single line. Same below.

charris commented on numpy/lib/npyio.py:
```diff
@@ -86,10 +86,17 @@ def __dir__(self):
+    
+    Allows for Zip64, and the `file` argument can accept file, str, or
+    pathlib.Path objects. `args` and `kwargs` are passed to the zipfile.ZipFile
+    constructor."""
```
Put a blank line and then the `"""` aligned with the first.

charris commented:
Looks good modulo two style nits.

charris commented:
Thanks @wackywendell .

ChrisBarker-NOAA commented:
Sorry to not look at the code -- but is this going to work in py2 with the pypi pathlib2 object? If so -- great!

charris commented:
No, it won't work with `pathlib2` because the module name differs. @wackywendell That could probably be fixed.

ChrisBarker-NOAA commented:
well, there is a new protocol being proposed -- something like a __fspath__ dunder method. So maybe just wait 'till that gets finalized.

charris commented:
@ChrisBarker-NOAA Probably a good idea. There is also `tofile` and `fromfile` to worry about, but they are at the C level and so it will be worth waiting to see what the final interface is.

eric-wieser commented:
That protocol has been finalized in [PEP 519](https://www.python.org/dev/peps/pep-0519/), and is live in python 3.6
